bug: Non-Deterministic Test ProxyProtocolIT.requiredModeShouldProxyKafkaRequestsAfterProxyHeader

A test failed non-deterministically in GitHub Actions:

## failed test

`io.kroxylicious.it.ProxyProtocolIT.requiredModeShouldProxyKafkaRequestsAfterProxyHeader`

https://github.com/kroxylicious/kroxylicious/actions/runs/24717567996/job/72315375616?pr=3748

## stack trace

```
[ERROR] io.kroxylicious.it.ProxyProtocolIT.requiredModeShouldProxyKafkaRequestsAfterProxyHeader -- Time elapsed: 5.234 s <<< FAILURE!
java.lang.AssertionError: 
[ApiVersions request should succeed after PROXY header] 
Expecting
  <CompletableFuture[Incomplete]>
to be completed within 5L Seconds.

exception caught while trying to get the future result: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
	at org.assertj.core.internal.Futures.assertSucceededWithin(Futures.java:104)
	at org.assertj.core.api.AbstractCompletableFutureAssert.internalSucceedsWithin(AbstractCompletableFutureAssert.java:475)
	at org.assertj.core.api.AbstractCompletableFutureAssert.succeedsWithin(AbstractCompletableFutureAssert.java:471)
	at io.kroxylicious.it.ProxyProtocolIT.requiredModeShouldProxyKafkaRequestsAfterProxyHeader(ProxyProtocolIT.java:131)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:786)
	at org.junit.platform.commons.support.ReflectionSupport.invokeMethod(ReflectionSupport.java:514)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:161)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:152)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:91)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:112)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:94)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:87)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$4(TestMethodTestDescriptor.java:221)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:217)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:159)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:70)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:157)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:147)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
```

The test expected an ApiVersions request to succeed after sending a PROXY header, but the CompletableFuture timed out after 5 seconds, suggesting a timing or race condition in the PROXY protocol handling that prevents the Kafka request from completing.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Format `<github issue/pr number>: <short description>`.
 
 ## SNAPSHOT
 
+* [#3757](https://github.com/kroxylicious/kroxylicious/issues/3757): fix(runtime): trigger read after HAProxy message to prevent deadlock when autoread disabled
 * [#1295](https://github.com/kroxylicious/kroxylicious/issues/1295): feat(aws-kms): add IRSA and EKS Pod Identity credential providers, and restructure AWS KMS credential configuration under a new `credentials` node (see [note](#note-021-aws-kms-credentials-restructure))
 * [#3745](https://github.com/kroxylicious/kroxylicious/issues/3745): fix(runtime): ensure async TransportSubjectBuilder callbacks execute on Netty event loop to prevent race conditions
 * [#3620](https://github.com/kroxylicious/kroxylicious/issues/3620): Removed Deprecated clientSaslAuthenticationSuccess from FilterContext

--- a/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/HaProxyMessageHandler.java
+++ b/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/HaProxyMessageHandler.java
@@ -46,5 +46,7 @@ public class HaProxyMessageHandler extends SimpleChannelInboundHandler<HAProxyMe
                 .addKeyValue("destinationPort", haProxyMessage.destinationPort())
                 .log("Received HAProxy message");
         kafkaSession.setHaProxyContext(HaProxyContext.from(haProxyMessage));
+        // Trigger next read to consume subsequent Kafka messages when autoread is disabled
+        ctx.channel().read();
     }
 }

--- a/kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/HaProxyMessageHandlerTest.java
+++ b/kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/HaProxyMessageHandlerTest.java
@@ -11,6 +11,7 @@ import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 
+import io.netty.channel.Channel;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.embedded.EmbeddedChannel;
 import io.netty.handler.codec.haproxy.HAProxyCommand;
@@ -27,6 +28,7 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoInteractions;
 import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
 class HaProxyMessageHandlerTest {
@@ -48,6 +50,9 @@ class HaProxyMessageHandlerTest {
     @Mock
     private ChannelHandlerContext ctx;
 
+    @Mock
+    private Channel channel;
+
     private HaProxyMessageHandler handler;
 
     @BeforeEach
@@ -57,6 +62,9 @@ class HaProxyMessageHandlerTest {
 
     @Test
     void shouldStoreHaProxyContextInSession() throws Exception {
+        // Given
+        when(ctx.channel()).thenReturn(channel);
+
         // When
         handler.channelRead(ctx, newHAProxyMessage());
 
@@ -66,20 +74,32 @@ class HaProxyMessageHandlerTest {
     }
 
     @Test
+    void shouldTriggerReadAfterProcessingHaProxyMessage() throws Exception {
+        // Given
+        when(ctx.channel()).thenReturn(channel);
+
+        // When
+        handler.channelRead(ctx, newHAProxyMessage());
+
+        // Then - channel.read() should be called to consume next message when autoread is disabled
+        verify(channel).read();
+    }
+
+    @Test
     void shouldReleaseHaProxyMessageAfterProcessing() {
         // Given
         HAProxyMessage message = newHAProxyMessage();
         assertThat(message.refCnt()).isEqualTo(1);
-        EmbeddedChannel channel = new EmbeddedChannel(handler);
+        EmbeddedChannel embeddedChannel = new EmbeddedChannel(handler);
 
         // When
-        channel.writeInbound(message);
+        embeddedChannel.writeInbound(message);
 
         // Then - SimpleChannelInboundHandler must release the ref-counted message
         assertThat(message.refCnt()).isZero();
-        assertThat((Object) channel.readInbound()).isNull();
+        assertThat((Object) embeddedChannel.readInbound()).isNull();
         verify(kafkaSession).setHaProxyContext(any(HaProxyContext.class));
-        channel.finishAndReleaseAll();
+        embeddedChannel.finishAndReleaseAll();
     }
 
     @Test

